ampd/src/handlers/starknet_verify_verifier_set.rs

```diff
@@ -0,0 +1,296 @@
+use crate::types::TMAddress;
+
+#[derive(Deserialize, Debug)]
+pub struct VerifierSetConfirmation {
+    pub message_id: FieldElementAndEventIndex,
+    pub verifier_set: VerifierSet,
+}
+
+#[derive(Deserialize, Debug)]
+#[try_from("wasm-verifier_set_poll_started")]
+struct PollStartedEvent {
+    poll_id: PollId,
+    source_gateway_address: String,
+    verifier_set: VerifierSetConfirmation,
+    participants: Vec<TMAddress>,
+    expires_at: u64,
+}
+
+pub struct Handler<C>
+where
+    C: StarknetClient + Send + Sync,
+{
+    verifier: TMAddress,
+    voting_verifier_contract: TMAddress,
+    rpc_client: C,
+    latest_block_height: Receiver<u64>,
+}
+
+impl<C> Handler<C>
+where
+    C: StarknetClient + Send + Sync,
+{
+    /// Handler for verifying verifier set updates from Starknet
+    ///
+    /// # Type Parameters
+    /// * `C` - A Starknet client type that implements the [`StarknetClient`] trait
+    pub fn new(
+        verifier: TMAddress,
+        voting_verifier_contract: TMAddress,
+        rpc_client: C,
+        latest_block_height: Receiver<u64>,
+    ) -> Self {
+        Self {
+            verifier,
+            voting_verifier_contract,
+            rpc_client,
+            latest_block_height,
+        }
+    }
+
+    fn vote_msg(&self, poll_id: PollId, vote: Vote) -> MsgExecuteContract {
+        MsgExecuteContract {
+            sender: self.verifier.as_ref().clone(),
+            contract: self.voting_verifier_contract.as_ref().clone(),
+            msg: serde_json::to_vec(&ExecuteMsg::Vote {
+                poll_id,
+                votes: vec![vote],
+            })
+            .expect("vote msg should serialize"),
+            funds: vec![],
+        }
+    }
+}
+
+#[async_trait]
+impl<C> EventHandler for Handler<C>
+where
+    C: StarknetClient + Send + Sync + 'static,
+{
+    type Err = Error;
+
+    async fn handle(&self, event: &Event) -> error_stack::Result<Vec<Any>, Self::Err> {
+        if !event.is_from_contract(self.voting_verifier_contract.as_ref()) {
+            return Ok(vec![]);
+        }
+
+        let PollStartedEvent {
+            poll_id,
+            source_gateway_address,
+            verifier_set,
+            expires_at,
+            participants,
+        } = match event.try_into() as error_stack::Result<_, _> {
+            Err(report) if matches!(report.current_context(), EventTypeMismatch(_)) => {
+                return Ok(vec![])
+            }
+            event => event.change_context(Error::DeserializeEvent)?,
+        };
+
+        if !participants.contains(&self.verifier) {
+            return Ok(vec![]);
+        }
+
+        if *self.latest_block_height.borrow() >= expires_at {
+            info!(poll_id = poll_id.to_string(), "skipping expired poll");
+            return Ok(vec![]);
+        }
+
+        let transaction_response = self
+            .rpc_client
+            .get_event_by_hash_signers_rotated(verifier_set.message_id.tx_hash.clone())
```

Comment:
correct me if I'm wrong, but it looks like the event index of the message id is being discarded? So one could put any value there, as long as the tx hash is correct. Is that right? If so, we should change that and also verify the event index

Reply:
We reworked it to also accept the `message_id` and get the event by it's index.
You can check this [commit](https://github.com/axelarnetwork/axelar-amplifier/pull/731/commits/dcbb6d1883b192b65ed9ae7d500a77c880c7ff9a) for the change.